test/shopify_app/controller_concerns/token_exchange_test.rb

```diff
@@ -178,13 +183,48 @@ class TokenExchangeControllerTest < ActionController::TestCase
     end
   end
 
+  [
+    ShopifyAPI::Errors::InvalidJwtTokenError,
+    ShopifyAPI::Errors::CookieNotFoundError,
+  ].each do |invalid_shopify_id_token_error|
+    test "Redirects to bounce page if Shopify ID token is invalid with #{invalid_shopify_id_token_error}" do
+      ShopifyApp.configuration.root_url = "/my-root"
+      ShopifyAPI::Utils::SessionUtils.stubs(:current_session_id).raises(invalid_shopify_id_token_error)
+      request.headers["HTTP_AUTHORIZATION"] = nil
+
+      params = { shop: @shop, my_param: "for-keeps", id_token: "dont-include-this-id-token" }
+      expected_redirect_url = "/my-root/patch_shopify_id_token?my_param=for-keeps&shop=my-shop.myshopify.com"
+      expected_redirect_url += "&shopify-reload=%2Freloaded_path%3Fmy_param%3Dfor-keeps%26shop%3Dmy-shop.myshopify.com"
```

Comment:
Nit: maybe we could URL encode the param for better readability of the test?
```suggestion
      reload_url = CGI.escape("/reloaded_path?my_param=for-keeps&shop=#{@shop}")
      expected_redirect_url = "/my-root/patch_shopify_id_token?my_param=for-keeps&shop=my-shop.myshopify.com&shopify-reload=#{reload_url}"
```